standalone/prouter/src/i2pd.rs

```diff
@@ -0,0 +1,483 @@
+use anyhow::{anyhow, Result};
+use log::{error, info};
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::path::PathBuf;
+
+#[allow(non_upper_case_globals)]
+#[allow(non_camel_case_types)]
+#[allow(non_snake_case)]
+include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+
+enum PRouterSigningKeyType {
+    SigningKeyTypeDSASha1 = 0,
+    SigningKeyTypeECDSASha256P256,
+    SigningKeyTypeECDSASha384P384,
+    SigningKeyTypeECDSASha512P521,
+    SigningKeyTypeRSASha2562048,
+    SigningKeyTypeRSASha3843072,
+    SigningKeyTypeRSASha5124096,
+    SigningKeyTypeEDDSASha512Ed25519,
+    SigningKeyTypeEDDSASha512Ed25519ph,
+    SigningKeyTypeGOSTR3410CryptoProAGostr3411256,
+    SigningKeyTypeGOSTR3410Tc26A512Gostr3411512,
+    SigningKeyTypeREDDSASha512Ed25519,
+}
+
+enum PRouterCryptoKeyType {
+    CryptoKeyTypeElgamal = 0,
+    CryptoKeyTypeECIESP256Sha256Aes256cbc = 1,
+    CryptoKeyTypeECIESX25519Aead = 4,
+    CryptoKeyTypeECIESP256Sha256Aes256cbcTest = 65280,
+    CryptoKeyTypeECIESGOSTR3410CryptoProASha256Aes256cbc = 65281,
+}
+
+#[derive(Debug)]
+pub struct TunnelInfo(pub String, pub String);
+
+#[derive(Debug, Clone)]
+pub struct I2pd {
+    argc: i32,
+    argv: HashMap<String, String>,
+    app_name: String,
+    is_running: bool,
+}
+
+// Here `phala_i2p_key` is the private key derived from the pRuntime private key with `PhalaI2PKey` info.
+// check `derive_phala_i2p_key()` method in `prpc_service.rs`
+pub fn generate_ident_to_file(
+    abs_datadir: &PathBuf,
+    filename: String,
+    phala_i2p_key: Vec<u8>,
+) -> Result<String> {
+    let mut keyfile_path = abs_datadir.clone();
+    keyfile_path.push(filename);
+
+    let c_sk: *const c_char = phala_i2p_key.as_ptr() as *const c_char;
+
+    let c_str_filename = CString::new(String::from(keyfile_path.to_string_lossy()))
+        .expect("String should be able to be converted into CString");
+    let c_filename: *const c_char = c_str_filename.as_ptr() as *const c_char;
+    let ptr_ident: *const c_char = unsafe {
+        C_GenerateIdentToFile(
+            c_filename,
+            c_sk,
+            PRouterSigningKeyType::SigningKeyTypeEDDSASha512Ed25519 as u16,
+            PRouterCryptoKeyType::CryptoKeyTypeElgamal as u16,
+        )
+    };
+
+    if ptr_ident.is_null() {
+        error!("The ident returned from C is corrupted");
+        return Err(anyhow!("The ident returned from C is corrupted"));
+    }
+
+    let c_str_ident = unsafe { CStr::from_ptr(ptr_ident) };
+    // Ok(c_str_ident.to_bytes().to_owned())
+    let ident: String = c_str_ident.to_str()?.to_owned();
+
+    Ok(ident)
+}
+
+fn get_client_tunnel_name_by_id(index: i32) -> Result<String> {
+    let ptr_name: *const c_char = unsafe { C_GetClientTunnelsName(index) };
```

Comment:
Since there are so many `*const c_char` returned from C APIs and not being freed, I assume they are pointing to some sort of static buffer on the C side. That means those C APIs are not thread-safe. So we could not wrap them into safe rust APIs. Because in the Rust world, if a user calls only safe APIs, it must not trigger any unsound behaviors. So we'd better re-do the CPI wrapping to export thread-safe APIs.

Reply:
Yes, there is a statically initialized memory containing returned values from C API. For now these C APIs are all for querying `i2pd` analysis data (e.g. connection speed, peers, etc), which is ensured to be single threaded. on the other hand, network packages are transmitted by a dedicated proxy so it should fully support multi-threaded sending/receiving.

The main problem for wrapping thread-safe APIs is that, C APIs return `string`, which is dynamically initialized on the C side. If we don't save the string to a static memory, it seems that the string will be automatically freed after the ffi call, and the received pointer in our rust code will therefore point to a meaningless memory. I am not sure how to handle this scenario, and I googled around and found no discussion or example about it as well. I would deeply appreciate it If you could provide some insights~

Overall, I think this problem should be find at least for now. However, I am open to discussion ;)

Reply:
## There 2 choices for C ffi:
1.
Declare:
```C
// copy the result into buf
int C_GetClientTunnelsName(int32_t index, char* buf, size_t buf_len);
```
Usage:
```rust
let mut buf[c_char; 64];
let ret = unsafe { C_GetClientTunnelsName(index, &mut buf, buf.len()) };
if ret != 0 {
   // error handling
}
```
2.
Declare:
```C
// Return a heap string which must be freed with C_FreeString
char* C_GetClientTunnelsName(int32_t index);
```
Usage:
```Rust
let ptr_name: *const c_char = unsafe { C_GetClientTunnelsName(index) };
if ptr_name.is_null() {
    // error handling and early return
}
do_somthing_with(ptr_name)
C_FreeString(ptr_name); // or libc::free(ptr_name) if the C side use the default allocator.
```

## However, I found the C side is wrapping C++ code?
You'd better take a look at [cxx](https://github.com/dtolnay/cxx) or [autocxx](https://github.com/google/autocxx)

Reply:
Fixed ;)